src/io/read_tree_genesis_hdf5.c

```diff
@@ -0,0 +1,1091 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "read_tree_genesis_hdf5.h"
+#include "hdf5_read_utils.h"
+#include "forest_utils.h"
+
+#include "../core_mymalloc.h"
+#include "../core_utils.h"
+
+
+#include <hdf5.h>
+
+/* Local enum for individual properties that are read in
+   from the Genesis hdf5 file*/
+enum GalaxyProperty
+{
+ head_enum = 0,
+ tail_enum = 1,
+ hosthaloid_enum,
+ m200c_enum,
+ /* m200b_enum, */
+ vmax_enum,
+ xc_enum,
+ yc_enum,
+ zc_enum,
+ vxc_enum,
+ vyc_enum,
+ vzc_enum,
+ len_enum,
+ mostboundid_enum,
+ lx_enum,
+ ly_enum,
+ lz_enum,
+ veldisp_enum,
+ num_galaxy_props /* should be the last one */
+};
+
+static char galaxy_property_names[num_galaxy_props][MAX_STRING_LEN];
+static int fix_flybys_genesis(struct halo_data *halos, const int64_t nhalos_last_snap);
+static void get_forest_metadata_filename(const char *forestfilename, const size_t stringlen, char *metadata_filename);
+
+#define CONVERSION_FACTOR_FOR_GENESIS_UNIQUE_INDEX      (1000000000000)
+#define CONVERT_HALOID_TO_SNAPSHOT(haloid)    (haloid / CONVERSION_FACTOR_FOR_GENESIS_UNIQUE_INDEX )
+#define CONVERT_HALOID_TO_INDEX(haloid)       ((haloid %  CONVERSION_FACTOR_FOR_GENESIS_UNIQUE_INDEX) - 1)
+#define CONVERT_SNAPSHOT_AND_INDEX_TO_HALOID(snap, index)    (snap*CONVERSION_FACTOR_FOR_GENESIS_UNIQUE_INDEX + index + 1)
+
+
+void get_forests_filename_genesis_hdf5(char *filename, const size_t len, const struct params *run_params)
+{
+    snprintf(filename, len - 1, "%s/%s%s", run_params->SimulationDir, run_params->TreeName, run_params->TreeExtension);
+}
+
+
+void get_forest_metadata_filename(const char *forestfilename, const size_t stringlen, char *metadata_filename)
+{
+    memmove(metadata_filename, forestfilename, stringlen);
+    metadata_filename[stringlen - 1] = '\0';
+
+    const char searchstring[] = {".hdf5"};
+    const size_t searchlen = strlen(searchstring);
+
+    const char replacestring[] = {".foreststats.hdf5"};
+    const size_t replacelen = strlen(replacestring);
+
+    const size_t currlen = strnlen(metadata_filename, stringlen);
+    const size_t required_len = currlen - searchlen + replacelen;
+    if(required_len >= stringlen) {
+        fprintf(stderr,"Error: Not enough memory reserved for 'metadata_filename' to correctly hold the file after "
+                                                "replacing string -- '%s' with (larger) string '%s'.\n"
+                "Please increase the size of 'metadata_filename' to be at least %zu\n",
+                searchstring, replacestring, required_len);
+        ABORT(INVALID_MEMORY_ACCESS_REQUESTED);
+    }
+
+    char *start = strstr(metadata_filename, searchstring);
+    strncpy(start, replacestring, replacelen);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120).  ](https://app.codacy.com/app/jacobseiler/sage-model/pullRequest?prid=4620074)